source/common/router/config_impl.cc

```diff
@@ -1087,7 +1088,25 @@ RouteConstSharedPtr RouteEntryImplBase::clusterEntry(const Http::HeaderMap& head
 RouteConstSharedPtr RouteEntryImplBase::pickWeightedCluster(const Http::HeaderMap& headers,
                                                             const uint64_t random_value,
                                                             const bool ignore_overflow) const {
-  const uint64_t selected_value = random_value % total_cluster_weight_;
+  uint64_t selected_value = random_value % total_cluster_weight_;
+  // Override the selected value with provided random value if related name is specified in the
+  // header.
+  if (!random_value_name_.empty()) {
+    const auto header_value = headers.get(Envoy::Http::LowerCaseString(random_value_name_));
+    if (!header_value.empty()) {
+      ASSERT(header_value.size() == 1);
+      uint64_t random_value_from_header = 0;
+      if (absl::SimpleAtoi(header_value[0]->value().getStringView(), &random_value_from_header)) {
+        selected_value = random_value_from_header % total_cluster_weight_;
+      }
+    } else {
+      // Random value should be found here. But if it is not found due to some errors, log the
+      // information and fallback to the random value that is set by stream id.
+      ENVOY_LOG(warn, "The random value can not be found from the header and it will fall back to "
```

Comment:
I'm not sure we can log each of these, it could be a DOS vector since we'd log one log line per request. I'd suggest a stat to track this, and if we want a log line we'd want to use one of the rate limiting log macros

Reply:
It is a good point about potential DOS vector. So, I have changed the logging level to `debug`, which matches other logging in this file.  `stat` might be a little bit overkill here for (but it could be a future improvement if needed as this feature evolves).